Handle Runnables in addition to Callables

rhuffman opened the pull request:
Added new method to Retryer:

```java
public void run(Runnable runnable) throws RetryException, InterruptedException
```
Changed the call method to throw RetryException and InterruptedException instead of Exception.

There was a lot of work on this branch attempting to have call and run methods that did not throw checked exceptions. But in the end most of those changes were rejected. InterruptedExceptions have to be dealt with, and were dealt with poorly in the original library. Now InterruptedExceptions from the BlockingStrategy are simply propagated rather than hoping the caller knows enough to check the interrupted state on the thread. And since InterruptedExceptions have to be propagated, there is no way to have a call or run method that does not at least throws that checked exception.

This PR also includes an update to JUnit 5. Currently only the RetryerTest uses the Jupiter API: all other tests are still written in JUnit 4.

rhuffman committed:
Refactor callables in RetryerTest

rhuffman committed:
Stop using ExecutionException to wrap exceptions thrown by Callables

== DETAILS

There does not seem to be any point to wrapping the Exceptions thrown
by the retried Callables in an ExcutionException. In fact, it just
makes it more complicated to use the Retryer. So I'm changing that
behavior.

Also, the Attempt.getExceptionCause method has been renamed to just
getException.

Fixes #22

rhuffman committed:
Add support for retrying a Runnable

rhuffman committed:
Make tests more succint

rhuffman committed:
Add JUnit 5 and parameterize the Retryer test

rhuffman committed:
Add test for all retries failing

rhuffman committed:
Change behavior when retried method always throws

== DETAILS

Prior to this change, a Retryer that was stopping would throw
a RetryException. However, that typically means the user just
has to unwrap the RetryException to get at the real cause.

Therefore, I'm changing the behavior to simply propagate the retried
method's Exception when the Retryer stops.

Fixes #23

rhuffman committed:
Propagate InterruptedException from the call method

== DETAILS

Wrapping exceptions in a RetryException just makes the caller's life
more difficult. Sure, it would give them the last attempt number, but
if they need that they can use a listener.

Fixes #24

rhuffman committed:
Add Retryer methods callUnchecked and runUnchecked

This included some refactoring in the call method.

dbyron0 commented on src/main/java/com/github/rholder/retry/Retryer.java:
```diff
@@ -84,69 +85,109 @@
                 }
+                Throwable throwable = attempt.getException();
+                if (throwable instanceof Exception) {
+                    throw (Exception) throwable;
```
Isn't this is behavior change for InterruptedException?  I'm not 100% sure, but by catching Throwable in makeAttempt, aren't we responsible for calling Thread.currentThread().interrupt() in case we're really dealing with an InterruptedException?  I've stared at http://www.yegor256.com/2015/10/20/interrupted-exception.html for awhile and I think so, but could be convinced otherwise.

rhuffman replied:
You are correct. I was going to add some tests for InterruptedExceptions before merging. I like to think I would have caught it, but maybe not. Thank you.

It makes me want to rethink this approach. The original guava-retrying also wrapped InterruptedExceptions, but it correctly interrupted the threads. However, I'm not sure throwing a non-InterruptedException is a good idea. It seems to me callers are very unlikely to handle it correctly.

I wanted the unchecked methods to make it easier to retry with lambdas that don't throw exceptions. But given that the retryer sleeps and thus can throw Interrupted, maybe it should just propagate InterruptedException. That will make it explicit to the caller. Of course, that means giving up the unchecked methods.

So perhaps I should rip out the unchecked methods again (bummer), then change the signatures of call and run to this:

```java
public <T> T call(Callable<T> callable) throws RetryException, InterruptedException
public void run(Runnable runnable) throws InterruptedException
```

Also, to get rid of the generic "throws Exception" in call, I should return to wrapping any exception that stops the attempt to be wrapped in RetryException. It's disappointing I can't make using lambdas easier, but I don't see any other way that is safe.

@dbyron0, what do you think?

dbyron0 replied:
I think ditching ExecutionException is good progress.  It would clean up my code.  I think this means the change to what RetryException means sticks around as well.  Being able to get rid of `throws Exception` seems good too.

I haven't totally digested the unchecked methods, but after a quick look I bet I stick with the `call` method.

Sorry if this isn't super helpful.  I bet tests with InterruptedException go a long way to working this out.

rhuffman replied:
Thanks. I'll go with the plan outlined above for now.

rhuffman committed:
Revert "Add Retryer methods callUnchecked and runUnchecked"

This reverts commit db6390c128b000fa17df8462a9be0ac3eacdff7e.

After considering how to deal with InterruptedExceptions (which can
happen because of the Thread.sleep method between attempts), I think
the unchecked methods are probably a bad idea.

The original guava-retrying library always wrapped the
InterruptedExceptions in a RetryException, and reset the interrupted
flag. However, I think that's probably not a great idea either:
callers are unlikely to check that flag and finish up their threads
immediately.

So I am going to remove the uncheck methods. Then I am going to cause
the call and run methods propagate InterruptedExceptions.

rhuffman committed:
Ignore annoying .attach_pid files

rhuffman committed:
Apply JUnit 5 Gradle plugin

I realized the single JUnit 5 test, RetryerTest, was not being run
from the test task.

rhuffman committed:
Remove unchecked methods

== DETAILS

This removes the callUnchecked and runUnchecked methods. Considering
that callers need to deal with the InterruptedException that can be
thrown by the sleep between attempts, I'm just propagating the
InterruptedException. My attempt to have an method that could take a
lambada converted to a Runnable with no exception handling was
laudable, but misguided.

I also changed the behavior so the call and run methods never just
propagate the exception from the Callable or Runnable; they always use
a RetryException.

rhuffman committed:
Add tests for InterruptedException

rhuffman committed:
Merge branch 'rhuffman/remove-unchecked-methods' into rhuffman/handle-runnables

rhuffman committed:
Change name of rejectionPredicates field

rhuffman committed:
Refactor Retryer

== DETAILS

Attempting to make the call method a little cleaner.

Also change the attempt number in Attempt from long to int, resolving
issue #27.

rhuffman committed:
Make common base class for Attempt classes

rhuffman committed:
Get rid of superfluous classes of BaseAttempt

There's really no need for subclasses of the BaseAttempt class.  This
just combines both into a single class. And since there are no longer
multiple implementation of the Attempt interface, the implementation
was moved up into the interface and it was changed to a class.

rhuffman committed:
Merge branch 'rhuffman/refactor-Retryer' into rhuffman/handle-runnables

rhuffman committed:
Simplify the way attempt times are dealt with

dbyron0 commented on src/main/java/com/github/rholder/retry/Retryer.java:
```diff
@@ -47,101 +44,119 @@
         Preconditions.checkNotNull(waitStrategy, "waitStrategy may not be null");
         Preconditions.checkNotNull(blockStrategy, "blockStrategy may not be null");
-        Preconditions.checkNotNull(rejectionPredicates, "rejectionPredicates may not be null");
+        Preconditions.checkNotNull(retryPredictes, "rejectionPredicates may not be null");
```
Another high-value comment :) rejectionPredicates --> retryPredicates in the message.

dbyron0 commented on src/main/java/com/github/rholder/retry/Retryer.java:
```diff
@@ -47,101 +44,119 @@
 
     /**
-     * Executes the given callable. If the rejection predicate
+     * Executes the given callable, retrying if necessary. If the rejection predicate
```
s/rejection predicate/retry predicate

dbyron0 commented on src/main/java/com/github/rholder/retry/Retryer.java:
```diff
@@ -47,101 +44,119 @@
             @Nonnull WaitStrategy waitStrategy,
             @Nonnull BlockStrategy blockStrategy,
-            @Nonnull List<Predicate<Attempt<?>>> rejectionPredicates,
+            @Nonnull List<Predicate<Attempt<?>>> retryPredictes,
```
s/retryPredictes/retryPredicates

dbyron0 commented on src/main/java/com/github/rholder/retry/Retryer.java:
```diff
@@ -47,101 +44,119 @@
+    }
+
+    /**
+     * Executes the given runnable, retrying if necessary. If the rejection predicate
```
s/rejection predicate/retry predicate

dbyron0 commented on src/test/java/com/github/rholder/retry/RetryerBuilderTest.java:
```diff
@@ -140,12 +143,8 @@ public void testRetryIfException() throws ExecutionException, RetryException {
-            assertTrue(e.getLastFailedAttempt().getExceptionCause() instanceof IOException);
-            assertTrue(e.getCause() instanceof IOException);
+            fail("Exception expected");
+        } catch (RetryException ignored) {
```
In case this floats your boat, check out https://github.com/joel-costigliola/assertj-core.  I specifically like [catchThrowable](https://github.com/joel-costigliola/assertj-core/blob/master/src/main/java/org/assertj/core/api/Assertions.java#L1197).

rhuffman committed:
Change 'rejectionPredicate' to 'retryPredicate' everywhere

rhuffman committed:
Cleanup: Fix javadoc and move a method lower in the file
